core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -1593,6 +1593,24 @@ trait ZIOFunctions extends Serializable {
    * current fiber for this operation to return non-empty lists.
    */
   final def children: UIO[IndexedSeq[Fiber[_, _]]] = descriptor.flatMap(_.children)
+
+  def fromOptionM[A](uio: UIO[Option[A]]): IO[Unit, A] = uio.flatMap {
+    case Some(a) => ZIO.succeed(a)
+    case None    => ZIO.fail(())
+  }
+
+  final def fromEitherE[E, A](io: Either[E, A]): IO[E, A] =
+    effectTotal(io).flatMap(_.fold(ZIO.fail, ZIO.succeed))
+
+  def fromEitherM[E, A](io: IO[E, Either[E, A]]): IO[E, A] = io.foldM(ZIO.fail, ZIO.fromEitherE)
```

Comment:
Since this function can fail, we have to move it to a trait with "fallible" methods, namely: `trait ZIOFunctions`. You can place it right next to `fromEither`, which it will call. We just have to add a type bound to `E` to ensure it can only be called from within suitable companion objects (i.e. `IO` and `ZIO` and `Task` but not `UIO`, because it fails).

The end result will be something like:

```scala
def fromEitherM[E <: UpperE, A](io: IO[E, Either[E, A]]): IO[E, A] = 
  io.foldM(fail, fromEither)
```

But living inside trait `ZIOFunctions`. 

The type bound exists only to ensure that the function cannot be called from `UIO.fromEitherM`, because people expect `UIO` methods to return `UIO` and not `IO`.

Reply:
```scala 
def fromEitherM[E <: UpperE, A](io: IO[E, Either[E, A]]): IO[E, A] =
  io.foldM(fail, fromEither)
```
It doesn't compiling.  
 polymorphic expression cannot be instantiated to expected type.
found   : [E(in method fromEither) <: ZIOFunctions.this.UpperE, A(in method fromEither)](=> Either[E(in method fromEither),A(in method fromEither)]) => scalaz.zio.IO[E(in method fromEither),A(in method fromEither)]

Reply:
Sorry to repeat myself, but doesn't `absolve` already do that?